File: android/app/src/main/java/uk/eugnel/ks2spelling/ParentAccessPlugin.java
package uk.eugnel.ks2spelling;

import android.app.KeyguardManager;
import android.content.Context;
import android.os.Build;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricPrompt;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;
import org.json.JSONObject;

@CapacitorPlugin(name = "ParentAccess")
public final class ParentAccessPlugin extends Plugin {
    private static final int BIOMETRIC_AUTHENTICATORS =
        BiometricManager.Authenticators.BIOMETRIC_STRONG;
    private static final int DEVICE_OWNER_AUTHENTICATORS =
        BiometricManager.Authenticators.BIOMETRIC_STRONG
            | BiometricManager.Authenticators.DEVICE_CREDENTIAL;
    // Quick unlock and PIN bootstrap/recovery are one native authority surface:
    // two prompts may never race each other.
    private final AtomicBoolean authenticationInFlight = new AtomicBoolean(false);

    @PluginMethod
    public void getBiometricAvailability(PluginCall call) {
        if (!exactKeys(call.getData(), new HashSet<>())) {
            rejectBiometric(call);
            return;
        }
        boolean available = BiometricManager.from(getContext())
            .canAuthenticate(BIOMETRIC_AUTHENTICATORS)
            == BiometricManager.BIOMETRIC_SUCCESS;
        JSObject result = new JSObject();
        result.put("available", available);
        result.put("type", available ? "biometric" : "none");
        call.resolve(result);
    }

    @PluginMethod
    public void authenticateBiometric(PluginCall call) {
        String reason = requireReason(call);
        if (reason == null) {
            rejectBiometric(call);
            return;
        }
        if (
            BiometricManager.from(getContext())
                .canAuthenticate(BIOMETRIC_AUTHENTICATORS)
                != BiometricManager.BIOMETRIC_SUCCESS
        ) {
            rejectBiometric(call);
            return;
        }
        if (!beginAuthentication(call, true)) return;
        FragmentActivity activity = (FragmentActivity) getActivity();
        activity.runOnUiThread(
            () -> showPrompt(activity, call, reason, false)
        );
    }

    @PluginMethod
    public void getDeviceOwnerAuthenticationAvailability(PluginCall call) {
        if (!exactKeys(call.getData(), new HashSet<>())) {
            rejectDeviceOwner(call);
            return;
        }
        JSObject result = new JSObject();
        result.put("available", deviceOwnerAuthenticationAvailable());
        call.resolve(result);
    }

    @PluginMethod
    public void authenticateDeviceOwner(PluginCall call) {
        String reason = requireReason(call);
        if (reason == null || !deviceOwnerAuthenticationAvailable()) {
            rejectDeviceOwner(call);
            return;
        }
        if (!beginAuthentication(call, false)) return;
        FragmentActivity activity = (FragmentActivity) getActivity();
        activity.runOnUiThread(
            () -> showPrompt(activity, call, reason, true)
        );
    }

    private boolean beginAuthentication(
        PluginCall call,
        boolean biometricOnly
    ) {
        if (!authenticationInFlight.compareAndSet(false, true)) {
            if (biometricOnly) rejectBiometric(call);
            else rejectDeviceOwner(call);
            return false;
        }
        if (!(getActivity() instanceof FragmentActivity)) {
            authenticationInFlight.set(false);
            if (biometricOnly) rejectBiometric(call);
            else rejectDeviceOwner(call);
            return false;
        }
        return true;
    }

    private boolean deviceOwnerAuthenticationAvailable() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            return BiometricManager.from(getContext())
                .canAuthenticate(DEVICE_OWNER_AUTHENTICATORS)
                == BiometricManager.BIOMETRIC_SUCCESS;
        }
        // AndroidX cannot reliably query DEVICE_CREDENTIAL through
        // canAuthenticate() before API 30. The platform's secure-keyguard bit
        // is the supported compatibility oracle for this repository's API-24
        // floor; the prompt below uses setDeviceCredentialAllowed(true).
        KeyguardManager keyguard = (KeyguardManager) getContext()
            .getSystemService(Context.KEYGUARD_SERVICE);
        return keyguard != null && keyguard.isDeviceSecure();
    }

    private void showPrompt(
        FragmentActivity activity,
        PluginCall call,
        String reason,
        boolean allowDeviceCredential
    ) {
        Executor executor = ContextCompat.getMainExecutor(getContext());
        BiometricPrompt prompt = new BiometricPrompt(
            activity,
            executor,
            new BiometricPrompt.AuthenticationCallback() {
                @Override
                public void onAuthenticationError(
                    int errorCode,
                    CharSequence errorText
                ) {
                    authenticationInFlight.set(false);
                    if (allowDeviceCredential) rejectDeviceOwner(call);
                    else rejectBiometric(call);
                }

                @Override
                public void onAuthenticationSucceeded(
                    BiometricPrompt.AuthenticationResult result
                ) {
                    authenticationInFlight.set(false);
                    JSObject response = new JSObject();
                    response.put("authenticated", true);
                    call.resolve(response);
                }
            }
        );
        BiometricPrompt.PromptInfo.Builder builder =
            new BiometricPrompt.PromptInfo.Builder()
                .setTitle(reason)
                .setSubtitle(
                    allowDeviceCredential
                        ? "Confirm the device owner"
                        : "Confirm that you are an adult"
                );
        if (allowDeviceCredential) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                builder.setAllowedAuthenticators(
                    DEVICE_OWNER_AUTHENTICATORS
                );
            } else {
                builder.setDeviceCredentialAllowed(true);
            }
        } else {
            builder
                .setAllowedAuthenticators(BIOMETRIC_AUTHENTICATORS)
                .setNegativeButtonText("Cancel");
        }
        prompt.authenticate(builder.build());
    }

    @Override
    protected void handleOnDestroy() {
        authenticationInFlight.set(false);
        super.handleOnDestroy();
    }

    private static String requireReason(PluginCall call) {
        if (!exactKeys(call.getData(), setOf("reason"))) return null;
        String reason = call.getString("reason");
        if (
            reason == null
                || reason.isEmpty()
                || reason.getBytes(StandardCharsets.UTF_8).length > 120
        ) {
            return null;
        }
        return reason;
    }

    private static boolean exactKeys(JSONObject value, Set<String> expected) {
        Set<String> actual = new HashSet<>();
        Iterator<String> keys = value.keys();
        while (keys.hasNext()) actual.add(keys.next());
        return actual.equals(expected) && value.length() == expected.size();
    }

    private static Set<String> setOf(String... values) {
        return new HashSet<>(Arrays.asList(values));
    }

    private static void rejectBiometric(PluginCall call) {
        call.reject(
            "Parent biometric authentication rejected.",
            "PARENT_BIOMETRICS_REJECTED"
        );
    }

    private static void rejectDeviceOwner(PluginCall call) {
        call.reject(
            "Parent device-owner authentication rejected.",
            "PARENT_DEVICE_AUTHENTICATION_REJECTED"
        );
    }
}

File: ios/App/App/ParentAccessPlugin.swift
import Capacitor
import Foundation
import LocalAuthentication

@objc(ParentAccessPlugin)
public final class ParentAccessPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ParentAccessPlugin"
    public let jsName = "ParentAccess"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(
            name: "getBiometricAvailability",
            returnType: CAPPluginReturnPromise
        ),
        CAPPluginMethod(
            name: "authenticateBiometric",
            returnType: CAPPluginReturnPromise
        ),
        CAPPluginMethod(
            name: "getDeviceOwnerAuthenticationAvailability",
            returnType: CAPPluginReturnPromise
        ),
        CAPPluginMethod(
            name: "authenticateDeviceOwner",
            returnType: CAPPluginReturnPromise
        )
    ]

    // Both native routes are one authority surface. A quick-unlock prompt and
    // a PIN bootstrap/recovery prompt may never overlap and race two calls.
    private var activeContext: LAContext?

    @objc public func getBiometricAvailability(_ call: CAPPluginCall) {
        guard call.options.keys.isEmpty else {
            rejectBiometric(call)
            return
        }
        let context = LAContext()
        var evaluationError: NSError?
        let available = context.canEvaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            error: &evaluationError
        )
        let type: String
        if available && context.biometryType == .faceID {
            type = "face"
        } else if available && context.biometryType == .touchID {
            type = "fingerprint"
        } else {
            type = "none"
        }
        call.resolve([
            "available": available && type != "none",
            "type": type
        ])
    }

    @objc public func authenticateBiometric(_ call: CAPPluginCall) {
        guard let reason = requireReason(call) else {
            rejectBiometric(call)
            return
        }
        DispatchQueue.main.async {
            guard self.activeContext == nil else {
                self.rejectBiometric(call)
                return
            }
            let context = LAContext()
            var evaluationError: NSError?
            guard context.canEvaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                error: &evaluationError
            ) else {
                self.rejectBiometric(call)
                return
            }
            self.activeContext = context
            context.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: reason
            ) { success, _ in
                DispatchQueue.main.async {
                    guard self.activeContext === context else {
                        self.rejectBiometric(call)
                        return
                    }
                    self.activeContext = nil
                    if success {
                        call.resolve(["authenticated": true])
                    } else {
                        self.rejectBiometric(call)
                    }
                }
            }
        }
    }

    @objc public func getDeviceOwnerAuthenticationAvailability(
        _ call: CAPPluginCall
    ) {
        guard call.options.keys.isEmpty else {
            rejectDeviceOwner(call)
            return
        }
        let context = LAContext()
        var evaluationError: NSError?
        let available = context.canEvaluatePolicy(
            .deviceOwnerAuthentication,
            error: &evaluationError
        )
        call.resolve(["available": available])
    }

    @objc public func authenticateDeviceOwner(_ call: CAPPluginCall) {
        guard let reason = requireReason(call) else {
            rejectDeviceOwner(call)
            return
        }
        DispatchQueue.main.async {
            guard self.activeContext == nil else {
                self.rejectDeviceOwner(call)
                return
            }
            let context = LAContext()
            var evaluationError: NSError?
            guard context.canEvaluatePolicy(
                .deviceOwnerAuthentication,
                error: &evaluationError
            ) else {
                self.rejectDeviceOwner(call)
                return
            }
            self.activeContext = context
            context.evaluatePolicy(
                .deviceOwnerAuthentication,
                localizedReason: reason
            ) { success, _ in
                DispatchQueue.main.async {
                    guard self.activeContext === context else {
                        self.rejectDeviceOwner(call)
                        return
                    }
                    self.activeContext = nil
                    if success {
                        call.resolve(["authenticated": true])
                    } else {
                        self.rejectDeviceOwner(call)
                    }
                }
            }
        }
    }

    private func requireReason(_ call: CAPPluginCall) -> String? {
        guard requireKeys(call, exactly: ["reason"]),
              let reason = call.getString("reason"),
              let length = reason.data(using: .utf8)?.count,
              length > 0,
              length <= 120 else {
            return nil
        }
        return reason
    }

    private func requireKeys(
        _ call: CAPPluginCall,
        exactly expected: Set<String>
    ) -> Bool {
        let keys = Set(call.options.keys.compactMap { $0 as? String })
        return keys == expected && call.options.keys.count == keys.count
    }

    private func rejectBiometric(_ call: CAPPluginCall) {
        call.reject(
            "Parent biometric authentication rejected.",
            "PARENT_BIOMETRICS_REJECTED"
        )
    }

    private func rejectDeviceOwner(_ call: CAPPluginCall) {
        call.reject(
            "Parent device-owner authentication rejected.",
            "PARENT_DEVICE_AUTHENTICATION_REJECTED"
        )
    }
}
